packages/recommend-react/src/useSafeEffect.ts

```diff
@@ -0,0 +1,55 @@
+import { useEffect, useRef } from 'react';
+
+type Props = Record<string, any>;
+type AlteredProps<TProps> = Partial<{ [TProp in keyof TProps]: any }>;
+
+function isPrimitive(obj: any) {
+  return Object(obj) !== obj;
+}
+
+function separatePrimitives<TProps>(props: TProps) {
+  const primitives = Object.entries(props).reduce<TProps>((prev, [k, v]) => {
+    if (isPrimitive(v)) {
+      return {
+        ...prev,
+        [k]: v,
+      };
+    }
+
+    return prev;
+  }, {} as TProps);
+  const nonPrimitives = Object.entries(props).reduce<TProps>((prev, [k, v]) => {
+    if (!isPrimitive(v)) {
+      return {
+        ...prev,
+        [k]: v,
+      };
+    }
+
+    return prev;
+  }, {} as TProps);
```

Comment:
these two could be combined in a single reduce if the accumulator was { primitives, nonPrimitives }

Reply:
Oh right! I mostly focused on make it work, hence the draft, but I'll update it thanks